¡Hola Jorge! Como te va?

Mi nombre es Facundo Lozano! Un gusto conocerte, seré tu revisor en este proyecto.

A continuación un poco sobre la modalidad de revisión que usaremos:

Cuando enccuentro un error por primera vez, simplemente lo señalaré, te dejaré encontrarlo y arreglarlo tú cuenta. Además, a lo largo del texto iré haciendo algunas observaciones sobre mejora en tu código y también haré comentarios sobre tus percepciones sobre el tema. Pero si aún no puedes realizar esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a comentarios y discusiones sobre el tema.

Encontrará mis comentarios a continuación: **no los mueva, modifique ni elimine**.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como este:

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Exito. Todo se ha hecho de forma exitosa.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Observación. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos. 
</div>

Puedes responder utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>


# Descripcion del proyecto # 8

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1) </b> <a class="tocSkip"></a>

Buenas Jorge! Siempre me tomo este tiempo al inicio de tu proyecto para comentarte mis apreciaciones generales de esta iteración de tu entrega. 

Me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes, te deseo lo mejor y espero que consigas lograr tus objetivos. Personalmente me gusta brindar el siguiente consejo, "Está bien equivocarse, es normal y es lo mejor que te puede pasar. Aprendemos de los errores y eso te hará mejor programando ya que podrás descubrir cosas a medida que avances y son estas cosas las que te darán esa experiencia para ser mejor como  Data Scientist"

Ahora si yendo a esta notebook. Lo he dicho al final del proyecto pero lo resalto aquí nuevamente, tu proyecto está muy bien resuelto Jorge, resalta capacidad y comprensión de todas las herrramientas, como a la vez esta ordenado y es sencillo de seguir, felictiaciones!

Este proyecto está en condiciones de ser aprobado! Éxitos dentro de tu camino en el mundo de los datos!

Saludos Jorge!

La compañía móvil Megaline no está satisfecha al ver que muchos de sus clientes utilizan planes heredados. Quieren desarrollar un modelo que pueda analizar el comportamiento de los clientes y recomendar uno de los nuevos planes de Megaline: Smart o Ultra.

Tienes acceso a los datos de comportamiento de los suscriptores que ya se han cambiado a los planes nuevos (del proyecto del sprint de Análisis estadístico de datos). Para esta tarea de clasificación debes crear un modelo que escoja el plan correcto.

Como ya hiciste el paso de procesar los datos, puedes lanzarte directo a crear el modelo.
Desarrolla un modelo con la mayor exactitud posible. En este proyecto, el umbral de exactitud es 0.75. Usa el dataset para comprobar la exactitud.

## Instrucciones del proyecto

- Abre y examina el archivo de datos. Dirección al archivo:datasets/users_behavior.csv Descarga el dataset
- Segmenta los datos fuente en un conjunto de entrenamiento, uno de validación y uno de prueba.
- Investiga la calidad de diferentes modelos cambiando los hiperparámetros. Describe brevemente los hallazgos del estudio.
- Comprueba la calidad del modelo usando el conjunto de prueba.
- Tarea adicional: haz una prueba de cordura al modelo. Estos datos son más complejos que los que habías usado antes así que no será una tarea fácil. Más adelante lo veremos con más detalle.


Cada observación en el dataset contiene información del comportamiento mensual sobre un usuario. La información dada es la siguiente:

- сalls — número de llamadas,
- minutes — duración total de la llamada en minutos,
- messages — número de mensajes de texto,
- mb_used — Tráfico de Internet utilizado en MB,
- is_ultra — plan para el mes actual (Ultra - 1, Smart - 0).

## Evaluación del proyecto

Hemos definido los criterios de evaluación para el proyecto. Lee esto con atención antes de pasar al ejercicio. 
Esto es lo que los revisores buscarán cuando evalúen tu proyecto:

- ¿Cómo leíste los datos después de descargarlos?
- ¿Segmentaste correctamente los datos en conjuntos de entrenamiento, validación y prueba?
- ¿Cómo escogiste el tamaño de los conjuntos?
- ¿Evaluaste correctamente la calidad del modelo?
- ¿Qué modelos e hiperparámentros usaste?
- ¿Cuáles fueron tus hallazgos?
- ¿Probaste los modelos correctamente?
- ¿Cuál es tu puntuación de exactitud?
- ¿Te ceñiste a la estructura del proyecto y mantuviste limpio el código?

Tienes tus hojas informativas y los resúmenes de los capítulos así que ya puedes continuar con el proyecto.
¡Buena suerte!

### Importar librerias necesarias para el proyecto

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from IPython.display import HTML
from sklearn.model_selection import GridSearchCV

### Importar Archivos

In [11]:
df = pd.read_csv("https://practicum-content.s3.us-west-1.amazonaws.com/datasets/users_behavior.csv")

### Mostrar datos relevantes del dataframe

In [12]:
display(HTML('<hr>'))
display(HTML('<h1> Datos de comportamiento de usuarios '))
display(df.head())
display(df.info())
display(df.describe())

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


None

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)
    
</b> <a class="tocSkip"></a>

    
Excelente implementación de importaciones y carga de datos. A la vez excelente implementación de los métodos para observar la composición de los datos!

### Verificacion de datos nulos 

In [13]:
nulos = df.isnull().sum()
display(HTML('<h2> Verificacion de valores nulos'))
display(nulos)


calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

### Verificación de valores duplicados

In [14]:
duplicados = df.duplicated().sum()
display(HTML('<h2> Verificación de valores duplicados </h2>'))
display(f"Total de valores duplicados: {duplicados}")

display(HTML('<hr>'))

'Total de valores duplicados: 0'

<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)
    
</b> <a class="tocSkip"></a>

    
Y como siempre es útil corroborar los valores duplicados y ausentes, bien hecho, sigamos!

### Comentario sobre la exploracion inicial del archivo

Tras Realizar la exploracion inicial de datos se observo que no hay datos duplicados o nulos dentro del conjunto. Este conjunto de datos recopila informacion sobre el comportamiento de los usuarios, lo cuales engloban las llamadas, minutos utilizados, mensaje enviados y el uso de datos en MB

### Segmentacion de datos en conjuntos de entrenamiento, validacion y prueba

In [21]:
# Dividir los datos en características (X) y la variable objetivo (y)
features = df.drop('is_ultra', axis=1)
target = df['is_ultra']

# Dividir los datos en conjuntos de entrenamiento (70%), validación (15%), y prueba (15%)
X_train, X_temp, y_train, y_temp = train_test_split(features, target, test_size=0.3, random_state=12345)
X_valid, X_test, y_valid, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=12345)

# Muestra el tamaño de cada conjunto
print("Tamaño del conjunto de entrenamiento:", len(X_train))
print("Tamaño del conjunto de validación:", len(X_valid))
print("Tamaño del conjunto de prueba:", len(X_test))

Tamaño del conjunto de entrenamiento: 2249
Tamaño del conjunto de validación: 482
Tamaño del conjunto de prueba: 483


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)
    
</b> <a class="tocSkip"></a>

    
Nuevamente excelente! División perfecta de los datos en los 3 conjuntos correspondientes, bien hecho!

### Aplicar modelo Random Forest y evaluacion

In [22]:
#Definir la grilla de hiperparámetros para Random Forest
param_grid_rf = {
    'n_estimators': [100, 150, 200],
    'max_depth': [5, 10, None],
    'min_samples_leaf': [1, 4, 7],
    'min_samples_split': [2, 5, 10]
    }

# Crear el objeto GridSearchCV
grid_search_rf = GridSearchCV(estimator=RandomForestClassifier(random_state=12345), 
                              param_grid=param_grid_rf, 
                              cv=5, 
                              scoring='accuracy')

# Ajustar GridSearchCV
grid_search_rf.fit(X_train, y_train)

# Mejores hiperparámetros
print("Mejores hiperparámetros para Random Forest:", grid_search_rf.best_params_)

# Crear una nueva instancia del modelo con los mejores hiperparámetros
best_rf_model = RandomForestClassifier(n_estimators=grid_search_rf.best_params_['n_estimators'],
                                       max_depth=grid_search_rf.best_params_['max_depth'],
                                       min_samples_leaf=grid_search_rf.best_params_['min_samples_leaf'],
                                       min_samples_split=grid_search_rf.best_params_['min_samples_split'],
                                       random_state=12345)

# Entrenar el modelo con el conjunto de entrenamiento
best_rf_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba y evaluar
predictions_test_rf = best_rf_model.predict(X_test)
accuracy_test_rf = accuracy_score(y_test, predictions_test_rf)
print("Exactitud del modelo Random Forest en el conjunto de prueba:", accuracy_test_rf)

Mejores hiperparámetros para Random Forest: {'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}
Exactitud del modelo Random Forest en el conjunto de prueba: 0.8033126293995859


### Aplicar modelo Logistic Regression y evaluacion

In [17]:
# Definir la grilla de hiperparámetros para Regresión Logística
param_grid_lr = {
    'C': [0.01, 0.1, 1, 10],
    'penalty': ['l1', 'l2'],
    'solver': ['liblinear']
}

# Crear el objeto GridSearchCV
grid_search_lr = GridSearchCV(estimator=LogisticRegression(random_state=12345, max_iter=200), param_grid=param_grid_lr, cv=5, scoring='accuracy')

# Ajustar GridSearchCV
grid_search_lr.fit(X_train, y_train)

# Mejores hiperparámetros
print("Mejores hiperparámetros para Logistic Regression:", grid_search_lr.best_params_)

# Crear una nueva instancia del modelo con los mejores hiperparámetros

best_lr_model = LogisticRegression(C=grid_search_lr.best_params_['C'],
                                       penalty=grid_search_lr.best_params_['penalty'],
                                       solver=grid_search_lr.best_params_['solver'],
                                       random_state=12345)

# Entrenar el modelo con el conjunto de entramiento
best_lr_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba y evaluar
predictions_test_lr = best_lr_model.predict(X_test)
accuracy_test_lr = accuracy_score(y_test, predictions_test_lr)
print("Exactitud del modelo Logistic Regression en el conjunto de prueba:", accuracy_test_lr)



Mejores hiperparámetros para Logistic Regression: {'C': 1, 'penalty': 'l1', 'solver': 'liblinear'}
Exactitud del modelo Logistic Regression en el conjunto de prueba: 0.7391304347826086


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)
    
</b> <a class="tocSkip"></a>

    
Unos modelos perfectos Jorge! El procedimiento es muy completo teniendo en cuenta 2 tipos de modelos y utilizando técnicas como GridSearch para entcontrar los mejores hiperparametros posibles. A la vez un excelente enternamiento y puesta en común de resultados con las métricas correspondientes!

### Prueba de cordura

In [18]:
# Analiza la importancia de las características
rf_feature_importances = pd.DataFrame(best_rf_model.feature_importances_, index=X_train.columns, columns=['importance']).sort_values('importance', ascending=False)


# Obtener los coeficientes del modelo
lr_coefficients = best_lr_model.coef_[0]

# Crear un DataFrame para visualizar los coeficientes (importancia) de cada característica
lr_feature_importance = pd.DataFrame(lr_coefficients, 
                                     index=X_train.columns, 
                                     columns=['Coefficient']).sort_values('Coefficient', ascending=False)

display(HTML('<hr>'))
display(HTML('<h2> Random forest: '))
print("Importancia de características en Random Forest:\n", rf_feature_importances)
display(HTML('<hr>'))
display(HTML('<h2> Logistic Regression: '))
print("Importancia de las características en la Regresión Logística:\n", lr_feature_importance)



Importancia de características en Random Forest:
           importance
mb_used     0.367198
minutes     0.232389
calls       0.202064
messages    0.198348


Importancia de las características en la Regresión Logística:
           Coefficient
messages     0.008967
calls        0.002473
minutes      0.001163
mb_used      0.000040


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)
    
</b> <a class="tocSkip"></a>

Implementación perfecta del modelo de cordura, bien hecho!

<div class="alert alert-block alert-warning">

<b>Comentario del revisor. (Iteración 1)
    
</b> <a class="tocSkip"></a>

Recordemos que siempre es positivo dejar conclusiones finales Jorge, aparte de eso el proyecto a estado perfecto!

## Concluciones

En el proceso de segmentación de datos, se dividieron los datos en conjuntos de entrenamiento (70%), validación (15%) y prueba (15%) para asegurar una evaluación adecuada de los modelos. Las características (X) se separaron de la variable objetivo (y), y las divisiones fueron confirmadas por los tamaños respectivos: entrenamiento (2249), validación (482), y prueba (483).

Se aplicaron y evaluaron dos modelos de machine learning: Random Forest y Logistic Regression. El modelo Random Forest, optimizado con los hiperparámetros max_depth: 10, min_samples_leaf: 1, min_samples_split: 2 y n_estimators: 100, logró una exactitud del 80.33% en el conjunto de prueba. La importancia de las características indicó que mb_used (datos usados) era la más relevante, seguida por minutes (minutos), calls (llamadas) y messages (mensajes).

Por otro lado, el modelo de Regresión Logística, optimizado con los hiperparámetros C: 1, penalty: l1 y solver: liblinear, alcanzó una exactitud del 73.91% en el conjunto de prueba. La importancia de las características mostró que messages tenía el mayor coeficiente, seguido por calls, minutes y mb_used.

En resumen, el modelo Random Forest demostró un mejor rendimiento en términos de exactitud comparado con la Regresión Logística. Además, la importancia de las características fue más significativa y clara en el modelo Random Forest.